### ISSUE TITLE
PlexOnlineToken

### ISSUE THREAD

**Issue by drymek** [linuxserver/docker-plex#60]
I've successfully installed and run docker image, but always getting message: 
```
plex    | #####################################################
plex    | #   Login via the webui at http://<ip>:32400/web    #
plex    | #   and restart the docker, because there was no    #
plex    | #      plex token found in the preference file      #
plex    | #####################################################
```
The web UI says (after login):
`A server could not be found. Download and install Plex Media Server, and it will appear here.
Where is my server?`
and
`Install your new server`

Host OS: 
```
lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.1 LTS
Release:	16.04
Codename:	xenial
```

docker-compose.yml file:
```
version: '2'

services:
  plex:
    image: linuxserver/plex
    container_name: plex
    environment:
      - VERSION=latest
      - PUID=1000
      - PGID=1000
    volumes:
      - "../volumes/config:/config"
      - "../volumes/data:/data"
      - "../volumes/transcode:/transcode"
    ports:
      - "32400:32400"
      - "32400:32400/udp"
      - "32469:32469"
      - "32469:32469/udp"
      - "5353:5353/udp"
      - "1900:1900/udp"
```

run command:
```
docker-compose up
```

Docker output:
```
plex    | [s6-init] making user provided files available at /var/run/s6/etc...exited 0.
plex    | [s6-init] ensuring user provided files have correct perms...exited 0.
plex    | [fix-attrs.d] applying ownership & permissions fixes...
plex    | [fix-attrs.d] done.
plex    | [cont-init.d] executing container initialization scripts...
plex    | [cont-init.d] 10-adduser: executing...
plex    |
plex    | -------------------------------------
plex    |           _     _ _
plex    |          | |___| (_) ___
plex    |          | / __| | |/ _ \
plex    |          | \__ \ | | (_) |
plex    |          |_|___/ |_|\___/
plex    |                |_|
plex    |
plex    | Brought to you by linuxserver.io
plex    | We do accept donations at:
plex    | https://www.linuxserver.io/donations
plex    | -------------------------------------
plex    | GID/UID
plex    | -------------------------------------
plex    | User uid:    1000
plex    | User gid:    1000
plex    | -------------------------------------
plex    |
plex    | [cont-init.d] 10-adduser: exited 0.
plex    | [cont-init.d] 30-dbus: executing...
plex    | [cont-init.d] 30-dbus: exited 0.
plex    | [cont-init.d] 40-chown-files: executing...
plex    | [cont-init.d] 40-chown-files: exited 0.
plex    | [cont-init.d] 50-plex-update: executing...
plex    |
plex    |
plex    |
plex    | #####################################################
plex    | #   Login via the webui at http://<ip>:32400/web    #
plex    | #   and restart the docker, because there was no    #
plex    | #      plex token found in the preference file      #
plex    | #####################################################
plex    |
plex    |
plex    | [cont-init.d] 50-plex-update: exited 0.
plex    | [cont-init.d] done.
plex    | [services.d] starting services
plex    | Starting dbus-daemon
plex    | Starting Plex Media Server.
plex    | 6 3000 /config/Library/Application Support
plex    | 8192
plex    | [services.d] done.
plex    | Starting Avahi daemon
plex    | Found user 'avahi' (UID 106) and group 'avahi' (GID 107).
plex    | Successfully dropped root privileges.
plex    | avahi-daemon 0.6.32-rc starting up.
plex    | No service file found in /etc/avahi/services.
plex    | Joining mDNS multicast group on interface eth0.IPv6 with address fe80::42:acff:fe12:9.
plex    | New relevant interface eth0.IPv6 for mDNS.
plex    | Joining mDNS multicast group on interface eth0.IPv4 with address 172.18.0.9.
plex    | New relevant interface eth0.IPv4 for mDNS.
plex    | Network interface enumeration completed.
plex    | Registering new address record for fe80::42:acff:fe12:9 on eth0.*.
plex    | Registering new address record for 172.18.0.9 on eth0.IPv4.
plex    | Server startup complete. Host name is 3b60b1204411.local. Local service cookie is 2452984411.
```

config/Library/Application\ Support/Plex\ Media\ Server/Preferences.xml (not sure what is secret, so I will remove them)
```
<?xml version="1.0" encoding="utf-8"?>
<Preferences OldestPreviousVersion="1.1.4.2757-24ffd60" MachineIdentifier="VALUE REMOVED" ProcessedMachineIdentifier="VALUE REMOVED" AnonymousMachineIdentifier="VALUE REMOVED" agentAutoEnabled.com.plexapp.agents.lastfm.Artists.com.plexapp.agents.vevo="1" MetricsEpoch="1" GracenoteUser="VALUE REMOVED"/>
```
I have plex account but don't have paid subscription. Restarting docker doesn't change anything. I've removed VERSION env, but the result (except the message) is the same - no plex server

If you need any other information just let me know - I will do my best to help.

**Comment by sparklyballs** [comment]
Firstly i would try running the container as outlined in the README with standard docker commands (ie not using composer) and using net=host.

Need to use as standard a setup as possible before we can diagnose any potential issues.

**Comment by patsissons** [comment]
I can confirm this bug. Ran the creation commands nearly verbatim and same result, restart results in no upgrade and no PMS is available to connect.

```
docker create \
--name=plex \
--net=host \
-e VERSION=latest \
-e PUID=... -e PGID=... \
-e TZ=... \
-v plex_config:/config \
-v plex_transcode:/transcode \
--volume-driver=convoy-nfs
linuxserver/plex
```

```
10/2/2016 1:05:02 PM #####################################################
10/2/2016 1:05:02 PM #   Login via the webui at http://<ip>:32400/web    #
10/2/2016 1:05:02 PM #   and restart the docker, because there was no    #
10/2/2016 1:05:02 PM #      plex token found in the preference file      #
10/2/2016 1:05:02 PM #####################################################
```

I can confirm that the container is otherwise working and data is being persisted to the named volume (i.e. the `Preferences.xml` file exists, but sign of `PlexOnlineToken`).

**Comment by sparklyballs** [comment]
try refreshing your browser if plex says no server found.

**Comment by patsissons** [comment]
I have already done this. force refreshed, cleared cache, tried another browser and other devices. definitely not a caching issue.

**Comment by sparklyballs** [comment]
i have in the last ten minutes or so, ran 5 pulls of the images from the hub into fresh config file locations with no failure to populate the preference file.
once i had no server found, a browser refresh found the server.

**Comment by patsissons** [comment]
let me spin up a new docker host and see what happens

**Comment by sparklyballs** [comment]
time for one of the world famous sparklyballs videos.

https://www.youtube.com/watch?v=Z2k_A4lUPEs

**Comment by patsissons** [comment]
interesting.

The only thing that I can see that may affect the outcomes is that I am running this with `docker-engine` which runs containers under root, whereas it appears you are running your containers with your unelevated account. I'm not sure if that would have an effect on things, but otherwise I can run the exact same command as you *verbatim* and get the same result.

`docker run -d --name=plex --net=host -e VERSION=latest -e PGID=1000 -e PUID=1000 -v /tmp/plex:/config linuxserver/plex`
```
# docker ps -a | grep plex
250d1084dffb        linuxserver/plex                "/init"                  3 minutes ago       Up 3 minutes                                                             plex

# ls -lan /tmp/plex/Library/Application\ Support/Plex\ Media\ Server/Preferences.xml
-rw------- 1 1000 1000 599 Oct  2 14:11 /tmp/plex/Library/Application Support/Plex Media Server/Preferences.xml

# cat /tmp/plex/Library/Application\ Support/Plex\ Media\ Server/Preferences.xml
<?xml version="1.0" encoding="utf-8"?>
<Preferences OldestPreviousVersion="1.1.4.2757-24ffd60" MachineIdentifier="..." ProcessedMachineIdentifier="..." AnonymousMachineIdentifier="..." agentAutoE
nabled.com.plexapp.agents.lastfm.Artists.com.plexapp.agents.vevo="1" MetricsEpoch="1" GracenoteUser="..."/>
```

The web interface is not notifying me of an unclaimed media server on the network, this may be where things are breaking down...

**Comment by patsissons** [comment]
**SOLVED**!

I was using internal dns to load the web ui, this is where things were breaking down. As soon as i hit the site with the docker node's IP address it was detected as a new PMS host and perform the initial setup (which I assume is where the `PlexOnlineToken` comes from).

Once the initial setup was complete I was able to use the internal dns again without any issues.

*postmortem*: perhaps a `README` patch to clarify that you must do the initial setup via IP.

**Comment by felixsteghofer** [comment]
FYI
I also tried every possible configuration (including the pristine default one), no chance to get the token into the configuration. After some time I found that 404, plex is throwing for the following request:

`https://plex.tv/api/home?X-Plex-Product=Plex%20Web&X-Plex-Version=2.8.1&...<some-ids-of-<you>...&X-Plex-Token=<your-plex-token>`

I don't know the exact reason why plex blocks this and the german status code: 
`<Response code="404" status="Sie sind gerade kein Teil des Hauses"/>` 
is not really helpful either.

But just putting the token into your `Preferences.xml` works:
`
<?xml version="1.0" encoding="utf-8"?>
<Preferences .......
PlexOnlineToken=<your-plex-token> />`

I do not really know in which exact step this is failing and overall looks odd to me.
Maybe somebody can help?

**Comment by drymek** [comment]
@patsissons  My issue is not connected to DNS. I'm using IP. 
@felixsteghofer Where did you find that?

**Comment by patsissons** [comment]
@drymek The issue I had wasn't specifically related to DNS, but rather how Plex web is detecting the PMS. The issue is that it is not able to detect a new server running if the hostname or IP it is running on is different from the hostname or IP you access it from. If you are still not using **host** networking or DNS there isn't a way to get that token (AFAIK). Without DNS, you'll need to at least run Plex with **host** networking to get the server detected, this is because without **host** networking the server runs on the internal IPv4 address (usually a 10.x.x.x address) which you cannot access from your browser. Even if you map ports and do any required forwarding, it won't work because you hit the server from the wrong address (the docker host's IP address). With **host** networking, the container binds to the docker host address and therefore listens on it as well. Once you do the initial setup, just redeploy your container without host networking (if you are concerned about that part) and re-enable your port mappings and everything should work fine. If **host** networking is not possible or desired at all, then you have to tell the container to use a custom hostname, and then set up your local DNS to point the hostname to the docker host. This is what I ended up doing and did not need to use **host** networking.

**Comment by drymek** [comment]
host network works like a charm, but when I change it later to ports mapping, the server could not be discovered. 

For others: the compose file with host networking is:
```
version: '2'

services:
  plex:
    image: linuxserver/plex
    container_name: plex
    environment:
      - VERSION=latest
      - PUID=1000
      - PGID=1000
    volumes:
      - "../volumes/config:/config"
      - "../volumes/data:/data"
      - "../volumes/transcode:/transcode"
    network_mode: "host"
```

**Comment by felixsteghofer** [comment]
@drymek 
you can see the error in your browsers javascript developer console (or in the "network" tab of the developer tools)

@patsissons 
interesting, I have the exact same setup, a custom hostname that is proxied to the plex container (and assigned to the plex container) but I couldn't get it to work this way. Always got the 404 described before

**Comment by drymek** [comment]
My mistake - switching back to ports mapping works as well.

**Comment by drymek** [comment]
@felixsteghofer Yes, there is:
```
vendors.a6e3046….js:287 GET https://plex.tv/api/home?X-Plex-Product=Plex%20Web&X-Plex-Version=2.8.1&X-P…ce-Screen-Resolution=1395x370%2C1440x900&X-Plex-Token=<token> 404 (Not Found)
```

**Comment by ns-cweber** [comment]
For posterity, I was having the same issue because I was pointing my browser at the server's name from my /etc/hosts file, which would hit the web server, but present the same "server could not be found" error message described above. When I entered the IP address of the server into my browser, everything worked just fine.

**Comment by nesrine19992** [comment]
I want to  use Radius to authentificate the users plex 
can you help me please ?

**Comment by CHBMB** [comment]
@nesrine19992 That's not supported by us, so don't think the linuxserver team is ignoring you but we can't help with this, however if some other person chimes in that's great though.

**Comment by nesrine19992** [comment]
thank you

**Comment by sooonmitch** [comment]
> @drymek The issue I had wasn't specifically related to DNS, but rather how Plex web is detecting the PMS. The issue is that it is not able to detect a new server running if the hostname or IP it is running on is different from the hostname or IP you access it from. If you are still not using **host** networking or DNS there isn't a way to get that token (AFAIK). Without DNS, you'll need to at least run Plex with **host** networking to get the server detected, this is because without **host** networking the server runs on the internal IPv4 address (usually a 10.x.x.x address) which you cannot access from your browser. Even if you map ports and do any required forwarding, it won't work because you hit the server from the wrong address (the docker host's IP address). With **host** networking, the container binds to the docker host address and therefore listens on it as well. Once you do the initial setup, just redeploy your container without host networking (if you are concerned about that part) and re-enable your port mappings and everything should work fine. If **host** networking is not possible or desired at all, then you have to tell the container to use a custom hostname, and then set up your local DNS to point the hostname to the docker host. This is what I ended up doing and did not need to use **host** networking.

Is it possible that we can add this to the Readme? Ran into the same issue because I directly started changing my ports instead of starting in host mode. With a hour or so searching through the issues, I found this gem and it fixes it. This would be very handy to have documented.
